src/main/java/minijava/ir/NumberOfLocalVariablesVisitor.java

```diff
@@ -0,0 +1,120 @@
+package minijava.ir;
+
+import java.util.ArrayList;
+import java.util.List;
+import minijava.ast.Block;
+import minijava.ast.BlockStatement;
+import minijava.ast.Expression;
+import minijava.ast.Statement;
+
+public class NumberOfLocalVariablesVisitor
+    implements Block.Visitor<Integer>,
+        Expression.Visitor<Integer>,
+        BlockStatement.Visitor<Integer> {
+  @Override
+  public Integer visitBlock(Block that) {
+    int c = 0;
+    List<Integer> localVarNums = new ArrayList<>();
+    for (BlockStatement statement : that.statements) {
+      if (statement instanceof Block
+          || statement instanceof Statement.While
+          || statement instanceof Statement.If) {
+        /* Statements that open a new scope */
+        localVarNums.add(c + statement.acceptVisitor(this));
+      } else {
+        c += statement.acceptVisitor(this);
+      }
+    }
+    localVarNums.add(c);
+    return localVarNums.stream().max(Integer::compare).get();
+  }
+
+  @Override
+  public Integer visitEmpty(Statement.Empty that) {
+    return 0;
+  }
+
+  @Override
+  public Integer visitIf(Statement.If that) {
+    int elseNum = 0;
+    if (that.else_.isPresent()) {
+      elseNum = that.else_.get().acceptVisitor(this);
+    }
+    return Math.max(that.then.acceptVisitor(this), elseNum);
```

Comment:
Something that I can say after understanding Firm variables: we should not take the maximum here, but allocate a new variable index for each Localvarbinding we encounter. So this should be `+` instead of `max`.

I was under the false impression that firm variables were for stack allocation rather than for constructing SSA form, where taking the maximum would've made sense.